NsqSharp/Consumer.cs

```diff
@@ -1383,6 +1406,86 @@ private void redistributeRDY()
             }
         }
 
+        private void redistributeRDYForStarvedConnections(List<Conn> connections, int maxInFlight)
+        {
+            var activeConns = new List<Conn>();
+            var idleConns = new List<Conn>();
+
+            // get idle and active connections
+            // idle = RDY > 0 and last message received > LowRdyIdleTimeout
+            // active = all other connections
+            // if an idle connection exists or an active connection with RDY=0, we're going to try to redistribute
+            foreach (var c in connections)
+            {
+                var lastMsgDuration = DateTime.Now.Subtract(c.LastMessageTime);
+                long rdyCount = c.RDY;
+                if (rdyCount > 0 && lastMsgDuration > _config.LowRdyIdleTimeout)
+                {
+                    idleConns.Add(c);
+                    _needRdyRedistributed = 1;
+                }
+                else
+                {
+                    activeConns.Add(c);
+                    if (rdyCount == 0)
+                    {
+                        // a better solution would be to leave 1 RDY on idle connections.
+                        // this would require minimum headroom on max rdy = concurrent handlers + (number of connections - 1)
+                        _needRdyRedistributed = 1;
+                    }
+                }
+            }
+
+            // check another thread didn't beat us to it
+            if (Interlocked.CompareExchange(ref _needRdyRedistributed, value: 0, comparand: 1) != 1)
```

Comment:
Thanks, the comment isn't necessary (or accurate), I've removed it. It's the same check used in `redistributeRDY`, its purpose is compare-and-swap, not anything to do with threading.

I was thinking a low `RDYRedistributeInterval` could cause this but the `select` in `rdyLoop` will block until `redistributeRDY` completes. Nice catch.